alpenhorn/cli/file/list.py

```diff
@@ -0,0 +1,342 @@
+
+    With --not, files must be absent from all the listed NODEs and GROUPs.
+
+    With --not and --all, files must be absent from at least one of the NODEs or
+    GROUPs.
+
+    \b
+    Limit by file state
+    -------------------
+
+    To limit the list based on file state, you can use the state flags:
+    --missing, --corrupt, --suspect, --healthy.  In this case, you must specify
+    at least one location (with --node or --group) to indicate where the file
+    state should be checked.  If multiple state flags are used, files with any
+    one of the specified state will be listed.
+
+    The --not flag cannot be used with the state flags.
+
+    Note: the "missing" state does not mean simply "absent".  The state "missing"
+    is a special state tracked by alpenhorn in the data index for files which are
+    absent from a node/group when they are expected to be present (i.e they haven't
+    been released).  So, these two commands produce (potentially) different results:
+
+    \b
+    group list -not --node=NODE
+    group list --missing --node=NODE
+
+    The first lists all files absent from NODE.  The second lists only files which are
+    absent but not released from the NODE (i.e. files which have been deleted by a
+    third-party without alpenhorn's knowledge).
+
+    \b
+    Limit based on syncability
+    --------------------------
+
+    Using both "--from=SRC" and "--to=DEST" will limit the list to files which
+    can be synced (copied) from the Node SRC to the Group DEST.  Specifically,
+    this restricts the list to files which are healthy on SRC and absent from
+    DEST.
+
+    You cannot simultaneously limit both by file state and syncability.
+    """
+
+    # Usage checks.
+
+    # Find a state flag to complain about, if any
+    if corrupt:
+        state_flag = "corrupt"
+    elif missing:
+        state_flag = "missing"
+    elif healthy:
+        state_flag = "healthy"
+    elif suspect:
+        state_flag = "suspect"
+    else:
+        state_flag = None
+
+    # Can't use a state flag with --not
+    not_both(not_, "not", state_flag, state_flag)
+    # Must use --to and --from together
+    both_or_neither(from_, "from", to, "to")
+    # Can't use a state with --from or --to
+    # (checking --from is enough, given the previous test)
+    not_both(from_, "from", state_flag, state_flag)
+
+    # Can't use --not or --all if we haven't used --node or --group
+    if not_ or all_:
+        if not node and not group:
+            raise click.UsageError(
+                ("--all" if all_ else "--not")
+                + " cannot be used without --node or --group"
+            )
+
+    # Figure out the state constraint from the flags
+    state_expr = state_constraint(
+        corrupt=corrupt, healthy=healthy, missing=missing, suspect=suspect
+    )
+
+    # Do we want the intersection or union of the nodes and groups?
+    # This is just XOR of --all and --not
+    intersect = all_ ^ not_
```

Comment:
I think the subtlety of why the XOR of `--not` and `--all` determines whether we use union or intersection is the same difficulty that the user might have when they try to figure out if they need `--not` or `--not --all`.  So, this might be another argument to drop `--not` and replace it with `--absent-node` &c. as I suggest below.